lib/Sema/CSSolver.cpp

```diff
@@ -792,6 +792,7 @@ bool ConstraintSystem::Candidate::solve(
 
   // Allocate new constraint system for sub-expression.
   ConstraintSystem cs(TC, DC, None);
+  cs.baseCS = &BaseCS;
```

Comment:
I see good and bad aspects to using baseCS for this. On the one hand that's currently only used by the diagnostics to find a contextual type, so this is extending how this is used (and if we ever wanted to eliminate this concept of a "base CS" it adds a little more friction to doing so).

On the other hand, this does make it possible to track total time from the creation of the first constraint system that spawns the other constraint systems. One question is whether this could change existing behavior in a meaningful way, e.g. we call typeCheckExpression from CSApply(), and before we would not have cared how long we spend rewriting the expression, but now we may time out on one of those subexpressions due to having a live Timer for the original constraint system.

Perhaps the solution for that is to clear out the `Optional<ExpressionTimer>` in certain places? Or perhaps it's not worth worrying about since we already give a lot of latitude here with the large threshold?

Reply:
We don't have to necessarily attach the parent constraint system here but it made sense from `baseCS` field description, although it's currently only used for diagnostics it makes sense to re-use to for shrinking as well because it splits expression into sub-expressions. 

I think we should be tracking all of the time spent doing work for expression and its sub-expressions irregardless of what it was doing because the threshold itself is lush enough so if it's triggered there is definitely something wrong with expression.

Reply:
Seems reasonable.